pytorch_lightning/core/lightning.py

```diff
@@ -1562,6 +1572,16 @@ def optimizer_step(
         """
         optimizer.step(closure=optimizer_closure)
 
+        # TODO: Refactor the following code out of this file.
```

Comment:
Personally, I am not a huge fan of the comm hooks leaking within the Plugin. Ideally, there should be their own isolated components.

Reply:
Here is an example: https://github.com/PyTorchLightning/pytorch-lightning/blob/490c40a8be339afd08dcfdcdd658db6b0be4671b/pytorch_lightning/plugins/ddp_comm_hooks/allreduce_invalid.py

Reply:
Agree that the current implementation is not a very elegant way, and I also think that ideally `lightning.py` file should be untouched in this PR.

I had a discussion with @ananthsub yesterday, who would later let someone do some follow-up refactoring. Sorry that I am not familiar with lightning internals, and actually I am leaving Facebook soon, so I really hope that this PR can be unblocked, and some follow-up works including refactoring and experimentations can be done concurrently.

Thanks for the understanding!